src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs

```diff
@@ -230,6 +230,8 @@ rootValue is not null &&
                 }
                 catch
                 {
+                    // Reset the writer in exception cases as we don't want the writer.Dispose() call in the finally block to flush any pending bytes.
```

Comment:
If we only do this to prevent the `writer.Dispose()` from flushing, why not just collocate the two calls?